Can't call methods of private subclasses in a module

Fixed an issue in #5841 where private subclasses could not be accessed.
However, the error still occurs if the class belongs to a module.

### Environment
$ java --version
openjdk 11.0.4 2019-07-16
OpenJDK Runtime Environment (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3)
OpenJDK 64-Bit Server VM (build 11.0.4+11-post-Ubuntu-1ubuntu218.04.3, mixed mode, sharing)

jruby-complete-9.2.9.0-20190926.210253-53.jar

JavaTest.java
```java
package test;

public class JavaTest {
    public interface PublicInterface {
        public void publicMethod(Integer a);
        public void publicMethod(Long a);
    }

    public PublicInterface publicMember = new PrivateSubclass();

    private class PrivateSubclass implements PublicInterface {
        public void publicMethod(Integer a) {System.out.println("PrivateSubclass.publicMethod(Integer)");}
        public void publicMethod(Long a) {System.out.println("PrivateSubclass.publicMethod(Long)");}
    }
}
```

module-info.java
```java
module test {
    exports test;
}
```

JavaMain.java
```java
import test.JavaTest;
import org.jruby.embed.ScriptingContainer;
public class JavaMain {
  public static void main(String[] args) {
    new JavaTest().publicMember.publicMethod(1);
    new ScriptingContainer().runScriptlet("java_import 'test.JavaTest'; JavaTest.new.publicMember.publicMethod(1)");
    new ScriptingContainer().runScriptlet("java_import 'test.JavaTest'; JavaTest.new.publicMember.java_method(:publicMethod, [java.lang.Integer.java_class]).call(1)");
  }
}
```

### Expected Behavior
```
$ javac -d mods/test module-info.java JavaTest.java
$ javac --module-path mods --add-modules test -cp jruby-complete.jar JavaMain.java
$ java --module-path mods --add-modules test -cp .:jruby-complete.jar JavaMain
PrivateSubclass.publicMethod(Integer)
PrivateSubclass.publicMethod(Long)
PrivateSubclass.publicMethod(Integer)
```

### Actual Behavior
```
$ javac -d mods/test module-info.java JavaTest.java
$ javac --module-path mods --add-modules test -cp jruby-complete.jar JavaMain.java
$ java --module-path mods --add-modules test -cp .:jruby-complete.jar JavaMain
PrivateSubclass.publicMethod(Integer)
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (jruby-complete.jar) to method sun.nio.ch.NativeThread.signal(long)
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
TypeError: illegal access on 'publicMethod': class org.jruby.javasupport.JavaMethod cannot access a member of class test.JavaTest$PrivateSubclass (in module test) with modifiers "public"
  <main> at <script>:1
Exception in thread "main" org.jruby.embed.EvalFailedException: (TypeError) illegal access on 'publicMethod': class org.jruby.javasupport.JavaMethod cannot access a member of class test.JavaTest$PrivateSubclass (in module test) with modifiers "public"
        at org.jruby.embed.internal.EmbedEvalUnitImpl.run(EmbedEvalUnitImpl.java:131)
        at org.jruby.embed.ScriptingContainer.runUnit(ScriptingContainer.java:1295)
        at org.jruby.embed.ScriptingContainer.runScriptlet(ScriptingContainer.java:1288)
        at JavaMain.main(JavaMain.java:7)
Caused by: org.jruby.exceptions.TypeError: (TypeError) illegal access on 'publicMethod': class org.jruby.javasupport.JavaMethod cannot access a member of class test.JavaTest$PrivateSubclass (in module test) with modifiers "public"
        at RUBY.<main>(<script>:1)
```

--- a/core/src/main/java/org/jruby/java/proxies/JavaProxy.java
+++ b/core/src/main/java/org/jruby/java/proxies/JavaProxy.java
@@ -9,9 +9,11 @@ import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.text.MessageFormat;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.function.Predicate;
 
 import com.headius.backport9.modules.Module;
 import com.headius.backport9.modules.Modules;
@@ -39,6 +41,7 @@ import org.jruby.javasupport.JavaClass;
 import org.jruby.javasupport.JavaMethod;
 import org.jruby.javasupport.JavaObject;
 import org.jruby.javasupport.JavaUtil;
+import org.jruby.javasupport.binding.MethodGatherer;
 import org.jruby.runtime.Helpers;
 import org.jruby.runtime.Block;
 import org.jruby.runtime.ObjectAllocator;
@@ -443,26 +446,24 @@ public class JavaProxy extends RubyObject {
 
     private Method getMethod(ThreadContext context, String name, Class... argTypes) {
         Class<?> originalClass = getObject().getClass();
+        Method[] holder = {null};
 
-        try {
-            for (Class<?> clazz = originalClass; clazz != null; clazz = clazz.getSuperclass()) {
-                Module module = Modules.getModule(clazz);
-                Package pkg = clazz.getPackage();
-
-                // Default package cannot be used by modules
-                if (pkg != null) {
-                    if (!module.isExported(pkg.getName())) continue;
+        Predicate<Method[]> predicate = (classMethods) -> {
+            for (Method method : classMethods) {
+                if (method.getName().equals(name) && Arrays.equals(method.getParameterTypes(), argTypes)) {
+                    holder[0] = method;
+                    return false;
                 }
-
-                Method method = clazz.getMethod(name, argTypes);
-
-                if (!Modifier.isPublic(method.getModifiers())) continue;
-
-                return method;
             }
-        } catch (NoSuchMethodException nsme) {
-            throw JavaMethod.newMethodNotFoundError(context.runtime, originalClass, name + CodegenUtils.prettyParams(argTypes), name);
-        }
+
+            return true;
+        };
+
+        MethodGatherer.eachAccessibleMethod(
+                originalClass,
+                predicate, predicate);
+
+        if (holder[0] != null) return holder[0];
 
         throw JavaMethod.newMethodNotFoundError(context.runtime, originalClass, name + CodegenUtils.prettyParams(argTypes), name);
     }

--- a/core/src/main/java/org/jruby/javasupport/binding/MethodGatherer.java
+++ b/core/src/main/java/org/jruby/javasupport/binding/MethodGatherer.java
@@ -28,6 +28,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Function;
+import java.util.function.Predicate;
 
 import static org.jruby.runtime.Visibility.PUBLIC;
 
@@ -152,6 +153,17 @@ public class MethodGatherer {
     static Map<String, List<Method>> getMethods(final Class<?> javaClass) {
         HashMap<String, List<Method>> nameMethods = new HashMap<>(32);
 
+        eachAccessibleMethod(
+                javaClass,
+                (classMethods) -> { addNewMethods(nameMethods, classMethods, true); return true; },
+                (interfaceMethods) -> { addNewMethods(nameMethods, interfaceMethods, false); return true; });
+
+        return nameMethods;
+    }
+
+    public static void eachAccessibleMethod(final Class<?> javaClass, Predicate<Method[]> classProcessor, Predicate<Method[]> interfaceProcessor) {
+        HashMap<String, List<Method>> nameMethods = new HashMap<>(32);
+
         // we scan all superclasses, but avoid adding superclass methods with
         // same name+signature as subclass methods (see JRUBY-3130)
         for ( Class<?> klass = javaClass; klass != null; klass = klass.getSuperclass() ) {
@@ -163,10 +175,10 @@ public class MethodGatherer {
                     // and replacing child methods with equivalent parent methods
                     PartitionedMethods filteredMethods = FILTERED_DECLARED_METHODS.get(klass);
 
-                    addNewMethods(nameMethods, filteredMethods.instanceMethods, true);
+                    if (!classProcessor.test(filteredMethods.instanceMethods)) return;
 
                     if (klass == javaClass) {
-                        addNewMethods(nameMethods, filteredMethods.staticMethods, true);
+                        if (!classProcessor.test(filteredMethods.staticMethods)) return;
                     }
                 }
                 catch (SecurityException e) { /* ignored */ }
@@ -180,13 +192,11 @@ public class MethodGatherer {
                     // parent methods
                     PartitionedMethods filteredMethods = FILTERED_METHODS.get(iface);
 
-                    addNewMethods(nameMethods, filteredMethods.instanceMethods, false);
+                    if (!interfaceProcessor.test(filteredMethods.instanceMethods)) return;
                 }
                 catch (SecurityException e) { /* ignored */ }
             }
         }
-
-        return nameMethods;
     }
 
     private static boolean methodsAreEquivalent(Method child, Method parent) {

